NativePlatforms returns the architecture as 32 bit x86 when running on ARM64

[NativePlatforms.java#L13](https://github.com/wpilibsuite/native-utils/blob/main/ToolchainPlugin/src/main/java/edu/wpi/first/toolchain/NativePlatforms.java#L13)

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':extractTestJNI'.
> Could not resolve all files for configuration ':nativeDesktopZip'.
   > Could not find opencv-cpp-3.4.7-5-osxx86.zip (edu.wpi.first.thirdparty.frc2021.opencv:opencv-cpp:3.4.7-5).
     Searched in the following locations:
         https://frcmaven.wpi.edu/artifactory/release/edu/wpi/first/thirdparty/frc2021/opencv/opencv-cpp/3.4.7-5/opencv-cpp-3.4.7-5-osxx86.zip
```

[I first noticed this error nearly a year ago](https://github.com/wpilibsuite/allwpilib/issues/3070) when I started using an M1 Macbook Pro. What I noticed was, at the time, the build compiled correctly when running natively. That was because my native Java install was an x86 build of AdoptOpenJDK running in Rosetta 2. It also compiled properly in the WPILib Docker image because Docker automatically emulated it in x86. When I recently updated to using an arm64 build of OpenJDK, that bug that I first noticed in January came back. This time, I traced the cause of the error to `edu.wpi.first.toolchain.NativePlatforms`. `NativePlatforms` is referenced [in the WPILib Gradle extension](https://github.com/wpilibsuite/GradleRIO/blob/main/src/main/java/edu/wpi/first/gradlerio/wpi/WPIExtension.java#L54) which is then called in [the standard build.gradle generated by the wpilib project creator](https://github.com/khari05/FRC-docker-template/blob/master/build.gradle#L48) when downloading C++ dependencies during the Java build.

Temporary Fix:
- Comment out the lines in build.gradle where it downloads packages specific to the desktop platform

Possible Solutions:
- intentionally make the architecture return x86-64 when aarch64/arm64 is detected so that it successfully downloads a file to make gradle happy for now and just warn the user that it is not officially supported
- make arm64 builds of linux (or maybe only Ubuntu) return the platform as `linuxaarch64bionic`
- [add arm64 macOS as a build target?](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary)

Helpful Info:

```
% java --version
openjdk 11.0.12 2021-07-20
OpenJDK Runtime Environment Homebrew (build 11.0.12+0)
OpenJDK 64-Bit Server VM Homebrew (build 11.0.12+0, mixed mode)
```

```groovy
import org.gradle.internal.os.OperatingSystem

println OperatingSystem.current() // Mac OS X 12.0.1 aarch64
println OperatingSystem.current().isMacOsX() // true
println System.getProperty("os.arch") // aarch64
```

--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ targetCompatibility = 11
 
 allprojects {
     group = "edu.wpi.first"
-    version = "2023.0.4"
+    version = "2023.0.5"
 
     if (project.hasProperty('publishVersion')) {
         version = project.publishVersion

--- a/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java
+++ b/src/main/java/edu/wpi/first/nativeutils/WPINativeUtilsExtension.java
@@ -11,6 +11,8 @@ import org.gradle.api.ExtensiblePolymorphicDomainObjectContainer;
 import org.gradle.api.model.ObjectFactory;
 import org.gradle.api.provider.ListProperty;
 import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
 
 import edu.wpi.first.nativeutils.platforms.PlatformConfig;
 import edu.wpi.first.nativeutils.dependencies.AllPlatformsCombinedNativeDependency;
@@ -41,6 +43,7 @@ public class WPINativeUtilsExtension {
         public final List<String> unixWarningArgs = List.of("-Wall", "-Wextra");
         public final List<String> unixWarningsAsErrorsArgs = List.of("-Werror");
 
+        public final String unixRpathOriginArg = "-Wl,-rpath,'$ORIGIN'";
         public final String unixSymbolArg = "-g";
 
         public final List<String> linuxCrossCompilerArgs = List.of("-std=c++17", "-Wformat=2", "-pedantic",
@@ -71,7 +74,7 @@ public class WPINativeUtilsExtension {
         public final List<String> macReleaseCompilerArgs = List.of("-O2");
         public final List<String> macDebugCompilerArgs = List.of("-O0");
         public final List<String> macLinkerArgs = List.of("-framework", "CoreFoundation", "-framework", "AVFoundation",
-                "-framework", "Foundation", "-framework", "CoreMedia", "-framework", "CoreVideo");
+                "-framework", "Foundation", "-framework", "CoreMedia", "-framework", "CoreVideo", "-headerpad_max_install_names");
     }
 
     public static class Platforms {
@@ -156,11 +159,13 @@ public class WPINativeUtilsExtension {
     }
 
     private final ObjectFactory objects;
+    private final ProviderFactory provider;
 
     @Inject
-    public WPINativeUtilsExtension(NativeUtilsExtension nativeExt, ObjectFactory objects) {
+    public WPINativeUtilsExtension(NativeUtilsExtension nativeExt, ObjectFactory objects, ProviderFactory provider) {
         this.nativeExt = nativeExt;
         this.objects = objects;
+        this.provider = provider;
 
         this.platforms = objects.newInstance(Platforms.class);
         defaultArguments = objects.newInstance(DefaultArguments.class);
@@ -214,6 +219,12 @@ public class WPINativeUtilsExtension {
         public abstract Property<String> getImguiVersion();
 
         public abstract Property<String> getWpimathVersion();
+
+        public abstract Property<String> getOpencvYear();
+
+        public abstract Property<String> getImguiYear();
+
+        public abstract Property<String> getGoogleTestYear();
     }
 
     private void addPlatformReleaseSymbolGeneration(String platform) {
@@ -309,6 +320,26 @@ public class WPINativeUtilsExtension {
         }
     }
 
+    private void addPlatformRpathAsOrigin(String platform) {
+        PlatformConfig plat = unixPlatforms.get(platform);
+        if (plat != null) {
+            plat.getLinker().getArgs().add(defaultArguments.unixRpathOriginArg);
+            return;
+        }
+    }
+
+    public void addPlatformRpathAsOrigin(String... platforms) {
+        if (platforms.length == 0) {
+            for (String platform : this.platforms.allPlatforms) {
+                addPlatformRpathAsOrigin(platform);
+            }
+        } else {
+            for (String platform : platforms) {
+                addPlatformRpathAsOrigin(platform);
+            }
+        }
+    }
+
     private DependencyVersions dependencyVersions;
 
     private void registerStandardDependency(ExtensiblePolymorphicDomainObjectContainer<NativeDependency> configs,
@@ -337,8 +368,34 @@ public class WPINativeUtilsExtension {
         });
     }
 
+    private void registerStandardDependency(ExtensiblePolymorphicDomainObjectContainer<NativeDependency> configs,
+            String name, Provider<String> groupId, String artifactId, Property<String> version) {
+        configs.register(name + "_shared", WPISharedMavenDependency.class, c -> {
+            c.getGroupId().set(groupId);
+            c.getArtifactId().set(artifactId);
+            c.getHeaderClassifier().set("headers");
+            c.getSourceClassifier().set("sources");
+            c.getExt().set("zip");
+            c.getVersion().set(version);
+            c.getExtraSharedExcludes().add("**/*java*");
+            c.getExtraSharedExcludes().add("**/*jni*");
+            c.getTargetPlatforms().addAll(this.platforms.allPlatforms);
+        });
+        configs.register(name + "_static", WPIStaticMavenDependency.class, c -> {
+            c.getGroupId().set(groupId);
+            c.getArtifactId().set(artifactId);
+            c.getHeaderClassifier().set("headers");
+            c.getSourceClassifier().set("sources");
+            c.getExt().set("zip");
+            c.getVersion().set(version);
+            c.getExtraSharedExcludes().add("**/*java*");
+            c.getExtraSharedExcludes().add("**/*jni*");
+            c.getTargetPlatforms().addAll(this.platforms.allPlatforms);
+        });
+    }
+
     private void registerStaticOnlyStandardDependency(ExtensiblePolymorphicDomainObjectContainer<NativeDependency> configs,
-            String name, String groupId, String artifactId, Property<String> version) {
+            String name, Provider<String> groupId, String artifactId, Property<String> version) {
         configs.register(name + "_static", WPIStaticMavenDependency.class, c -> {
             c.getGroupId().set(groupId);
             c.getArtifactId().set(artifactId);
@@ -350,11 +407,16 @@ public class WPINativeUtilsExtension {
         });
     }
 
+    private static String defaultDependencyYear = "frc2023";
+
     public void configureDependencies(Action<DependencyVersions> dependencies) {
         if (dependencyVersions != null) {
             return;
         }
         dependencyVersions = objects.newInstance(DependencyVersions.class);
+        dependencyVersions.getGoogleTestYear().set(defaultDependencyYear);
+        dependencyVersions.getImguiYear().set(defaultDependencyYear);
+        dependencyVersions.getOpencvYear().set(defaultDependencyYear);
         dependencies.execute(dependencyVersions);
         ExtensiblePolymorphicDomainObjectContainer<NativeDependency> configs = nativeExt.getNativeDependencyContainer();
         configs.register("netcomm", WPISharedMavenDependency.class, c -> {
@@ -417,9 +479,13 @@ public class WPINativeUtilsExtension {
 
         registerStandardDependency(configs, "wpimath", "edu.wpi.first.wpimath", "wpimath-cpp", dependencyVersions.getWpimathVersion());
 
-        registerStandardDependency(configs, "opencv", "edu.wpi.first.thirdparty.frc2022.opencv", "opencv-cpp", dependencyVersions.getOpencvVersion());
-        registerStaticOnlyStandardDependency(configs, "googletest", "edu.wpi.first.thirdparty.frc2022", "googletest", dependencyVersions.getGoogleTestVersion());
-        registerStaticOnlyStandardDependency(configs, "imgui", "edu.wpi.first.thirdparty.frc2022", "imgui", dependencyVersions.getImguiVersion());
+        Provider<String> opencvYearGroup = provider.provider(() -> "edu.wpi.first.thirdparty." + dependencyVersions.getOpencvYear().get() + ".opencv");
+        Provider<String> imguiYearGroup = provider.provider(() -> "edu.wpi.first.thirdparty." + dependencyVersions.getImguiYear().get());
+        Provider<String> googleTestYearGroup = provider.provider(() -> "edu.wpi.first.thirdparty." + dependencyVersions.getGoogleTestYear().get());
+
+        registerStandardDependency(configs, "opencv", opencvYearGroup, "opencv-cpp", dependencyVersions.getOpencvVersion());
+        registerStaticOnlyStandardDependency(configs, "googletest", googleTestYearGroup, "googletest", dependencyVersions.getGoogleTestVersion());
+        registerStaticOnlyStandardDependency(configs, "imgui", imguiYearGroup, "imgui", dependencyVersions.getImguiVersion());
 
 
         configs.register("wpilib_jni", AllPlatformsCombinedNativeDependency.class, c -> {

